plugins/gateway/tyk/v3.2.2/plugin.go

```diff
@@ -62,72 +78,104 @@ func ResponseSendTelemetry(rw http.ResponseWriter, res *http.Response, req *http
 		return
 	}
 
-	apiClient := utils.NewAPIClient("apiclarity.apiclarity:9000")
+	apiClient := newAPIClient(host)
 	params := operations.NewPostTelemetryParams().WithBody(telemetry)
 
 	_, err = apiClient.Operations.PostTelemetry(params)
 	if err != nil {
 		logger.Errorf("Failed to post telemetry: %v", err)
 		return
 	}
-	B, _ := json.Marshal(telemetry)
-	logger.Infof("Telemetry has been sent: %s", B)
+	logger.Infof("Telemetry has been sent")
 }
 
 func createTelemetry(res *http.Response, req *http.Request) (*models.Telemetry, error) {
+	truncatedBodyReq := false
+	truncatedBodyRes := false
+
 	apiDefinition := ctx.GetDefinition(req)
 	if apiDefinition == nil {
-		return nil, fmt.Errorf("failed to get api definition in response")
+		return nil, fmt.Errorf("failed to get api definition")
 	}
 
 	host, port := getHostAndPortFromTargetURL(apiDefinition.Proxy.TargetURL)
-	var destinationNamespace string
-	sp := strings.Split(host, ".")
-	if len(sp) == 2 {
-		destinationNamespace = sp[0]
-	}
+	destinationNamespace := getDestinationNamespaceFromHost(host)
 
 	reqBody, err := ioutil.ReadAll(req.Body)
 	if err != nil {
 		return nil, fmt.Errorf("failed to read request body: %v", err)
 	}
+	if len(reqBody) > MaxBodySize {
+		truncatedBodyReq = true
+		reqBody = []byte{}
+	}
 
 	resBody, err := ioutil.ReadAll(res.Body)
 	if err != nil {
 		return nil, fmt.Errorf("failed to read response body: %v", err)
 	}
+	if len(resBody) > MaxBodySize {
+		truncatedBodyRes = true
+		resBody = []byte{}
+	}
+
+	pathAndQuery := getPathWithQuery(req.URL)
 
 	telemetry := models.Telemetry{
 		DestinationAddress:   ":" + port, // No destination ip for now
 		DestinationNamespace: destinationNamespace,
 		Request: &models.Request{
 			Common: &models.Common{
-				TruncatedBody: false,
+				TruncatedBody: truncatedBodyReq,
 				Body:          strfmt.Base64(reqBody),
 				Headers:       createHeaders(req.Header),
 				Version:       req.Proto,
 			},
 			Host:   host,
 			Method: req.Method,
-			Path:   req.URL.String(),
+			Path:   pathAndQuery,
 		},
-		RequestID: uuid.NewV4().String(),
+		RequestID: getRequestIDFromHeaders(req.Header),
 		Response: &models.Response{
 			Common: &models.Common{
-				TruncatedBody: false,
+				TruncatedBody: truncatedBodyRes,
 				Body:          strfmt.Base64(resBody),
 				Headers:       createHeaders(res.Header),
-				Version:       req.Proto,
+				Version:       res.Proto,
 			},
 			StatusCode: strconv.Itoa(res.StatusCode),
 		},
-		Scheme:        req.Proto,
+		Scheme:        "http",
 		SourceAddress: req.RemoteAddr,
 	}
 
 	return &telemetry, nil
 }
 
+func getRequestIDFromHeaders(reqHeaders http.Header) string {
+	if reqID, ok := reqHeaders[RequestIDHeaderKey]; ok {
```

Comment:
maybe we want to search the headers case insensitive?

Reply:
header keys in go are represented in canonical form:
https://golang.org/src/net/http/header.go